frontend/src/app/spot/components/breadcrumbs/breadcrumbs-content.ts

```diff
@@ -0,0 +1,39 @@
+// -- copyright
+// OpenProject is an open source project management software.
+// Copyright (C) 2012-2022 the OpenProject GmbH
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License version 3.
+//
+// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+// Copyright (C) 2006-2013 Jean-Philippe Lang
+// Copyright (C) 2010-2013 the ChiliProject Team
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See COPYRIGHT and LICENSE files for more details.
+//++
+
+export interface Breadcrumb {
+  text:string;
+  icon?:string;
+  navigate?:() => void;
+}
+
+export class BreadcrumbsContent {
```

Comment:
Please do this as an interface instead of a class. Using classes just makes it harder to serialize/deserialize from JSON.

Reply:
but how do I make an interface immutable? What is the effect you mention, that it is harder to serialize? does it take more time?

Reply:
> What is the effect you mention, that it is harder to serialize?

You can't run `JSON.stringify` on this data. It might feel weird to think about this, but this is stateful data and we try to keep all stateful data serializable. I understand it's not possible for the callbacks in this interface (but they're not even used anyway, so you might as well delete them instead of offering a "fake" API.)

Reply:
immutability is key for view models, as it prevents the parent from altering the data in it while it is already "in use" for the rendered component. this makes the view more stable and IMHO easier to use, as you have only one binding and to alter it, you have to create a new view model. This way, you can always predict, how the view will look afterwards, there is no "hidden" data, which was bound to the view in some other initialization within the parent.

the callbacks are in use now, the template's buttons have the click handler defined.

And why is it, you should not be able to run `JSON.stringify` on a js object created with a class constructor?
![image](https://user-images.githubusercontent.com/38206611/197725496-eb404306-f548-49f6-b81b-a343c7251ed7.png)

Reply:
discussion delayed to include other devs ;)